src/t_zset.c

```diff
@@ -666,54 +679,66 @@ int zslIsInLexRange(zskiplist *zsl, zlexrangespec *range) {
+/* Find the Nth node that is contained in the specified range. N should be 0-based.
+ * Negative N works for reversed order (-1 represents the last element). Returns
+ * NULL when no element is contained in the range. */
+zskiplistNode *zslNthInLexRange(zskiplist *zsl, zlexrangespec *range, long n) {
     zskiplistNode *x;
     int i;
+    long rank = 0;
 
     /* If everything is out of range, return early. */
     if (!zslIsInLexRange(zsl,range)) return NULL;
 
     x = zsl->header;
-    for (i = zsl->level-1; i >= 0; i--) {
-        /* Go forward while *OUT* of range. */
-        while (x->level[i].forward &&
-            !zslLexValueGteMin(x->level[i].forward->ele,range))
+    if (n >= 0) {
+        for (i = zsl->level - 1; i >= 0; i--) {
+            /* Go forward while *OUT* of range. */
+            while (x->level[i].forward && !zslLexValueGteMin(x->level[i].forward->ele, range)) {
+                /* Count the rank of the first element in range. */
+                rank += x->level[i].span;
                 x = x->level[i].forward;
-    }
-
-    /* This is an inner range, so the next node cannot be NULL. */
-    x = x->level[0].forward;
-    serverAssert(x != NULL);
-
-    /* Check if score <= max. */
-    if (!zslLexValueLteMax(x->ele,range)) return NULL;
-    return x;
-}
-
-/* Find the last node that is contained in the specified range.
- * Returns NULL when no element is contained in the range. */
-zskiplistNode *zslLastInLexRange(zskiplist *zsl, zlexrangespec *range) {
-    zskiplistNode *x;
-    int i;
-
-    /* If everything is out of range, return early. */
-    if (!zslIsInLexRange(zsl,range)) return NULL;
-
-    x = zsl->header;
-    for (i = zsl->level-1; i >= 0; i--) {
-        /* Go forward while *IN* range. */
-        while (x->level[i].forward &&
-            zslLexValueLteMax(x->level[i].forward->ele,range))
+            }
+        }
+        if (n < ZSKIPLIST_MAX_SEARCH) {
+            /* Use the skip list to reach the nth node in range. */
+            for (i = 0; i < n + 1; i++) { 
+                if (!x) {
+                    return NULL;
+                }
+                x = x->level[0].forward; 
+            }
+        } else {
+            /* If offset is big, we jump to the nth node by its rank from the start of skiplist. */
+            x = (unsigned long)(rank + n) < zsl->length ? zslGetElementByRank(zsl, rank + n + 1) : NULL;
+        } 
+        /* Check if score <= max. */
+        if (x && !zslLexValueLteMax(x->ele,range)) return NULL;
+    } else {      
+        for (i = zsl->level - 1; i >= 0; i--) {
+            /* Go forward while *IN* range. */
+            while (x->level[i].forward && zslLexValueLteMax(x->level[i].forward->ele, range)) {
+                /* Count the rank of the last element in range. */
+                rank += x->level[i].span;
                 x = x->level[i].forward;
+            }
+        }  
+        if (n > -ZSKIPLIST_MAX_SEARCH) {
```

Comment:
```suggestion
        if (n + 1 > -ZSKIPLIST_MAX_SEARCH) {
```
n is started from -1.